File: src/main/java/com/parse/MessageEncoder.java
package com.parse;

import com.entity.GameOutput;
import com.socket.ActionData;
import com.util.IOUtils;
import com.util.ZlibUtil;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandler;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelPromise;
import io.netty.handler.codec.MessageToByteEncoder;

@ChannelHandler.Sharable
public class MessageEncoder extends MessageToByteEncoder<ActionData<?>> {

    private static MessageEncoder encoder;
    public static MessageEncoder getInst() {
        if (encoder == null) {
            synchronized (MessageEncoder.class) {
                if (encoder == null) {
                    encoder = new MessageEncoder();
                }
            }
        }
        return encoder;
    }

    @Override
    protected void encode(ChannelHandlerContext ctx, ActionData msg, ByteBuf out) throws Exception {
        GameOutput gameOutput = new GameOutput();
        try {
            gameOutput.writeLong(System.currentTimeMillis());// 发送当前服务器的时间
            gameOutput.writeInt(msg.getAction());
            byte[] buf = msg.getBuf();
            buf = buf == null ? new byte[0] : buf;
            gameOutput.writeInt(buf.length);
            gameOutput.write(buf, 0, buf.length);
            byte[] bytes = gameOutput.toByteArray();
            // 加密
//			System.out.println(bytes.length);
            String str = IOUtils.getAes(ctx.channel()).encrypt(bytes);
            gameOutput.reset();
            bytes = str.getBytes();
            // 压缩
            bytes = ZlibUtil.compress(bytes);
            out.writeInt(bytes.length);
            out.writeBytes(bytes);
        } finally {
            gameOutput.close();
        }
    }

    @Override
    public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
        super.close(ctx, promise);
    }

}

File: src/main/kotlin/com/parse/MessageEncoder.kt
package com.parse

import com.entity.Output
import com.socket.ActionData
import com.util.IOUtils.aes
import com.util.PoolUtils.getObject
import com.util.PoolUtils.returnObject
import com.util.ZlibUtil.compress
import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandler.Sharable
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelPromise
import io.netty.handler.codec.MessageToByteEncoder

@Sharable
class MessageEncoder(
    /**
     * 是否加密
     */
    private var isEncrypt: Boolean = false
) : MessageToByteEncoder<ActionData<*>>() {
    @Throws(Exception::class)
    override fun encode(ctx: ChannelHandlerContext, msg: ActionData<*>, out: ByteBuf) {
        val gameOutput = getObject(Output::class.java)
        try {
            gameOutput.writeLong(System.currentTimeMillis()) // 发送当前服务器的时间
            gameOutput.writeInt(msg.action)
            var buf = msg.buf
            buf = buf ?: ByteArray(0)
            gameOutput.writeInt(buf.size)
            gameOutput.write(buf, 0, buf.size)
            var bytes = gameOutput.toByteArray()
            // 加密
//			System.out.println(bytes.length);
            val str = aes.encrypt(bytes)
            gameOutput.reset()
            bytes = str.toByteArray()
            // 压缩
            bytes = compress(bytes)
            out.writeInt(bytes.size)
            out.writeBytes(bytes)
        } finally {
            returnObject(gameOutput)
        }
    }

    override fun close(ctx: ChannelHandlerContext, promise: ChannelPromise) {
        super.close(ctx, promise)
    }

    companion object {

        val inst by lazy(LazyThreadSafetyMode.SYNCHRONIZED) { MessageEncoder() }
        fun inst(isEncrypt: Boolean): MessageEncoder {
            return inst.also { it.isEncrypt = isEncrypt }
        }

    }
}
